messages/event_manager_test.go

```diff
@@ -1,26 +1,212 @@
 package messages
 
 import (
+	"sync/atomic"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 
 	"github.com/0xPolygon/go-ibft/messages/proto"
 )
 
-func TestEventManager_SubscribeCancel(t *testing.T) {
-	t.Parallel()
-
-	numSubscriptions := 10
-	subscriptions := make([]*Subscription, numSubscriptions)
-	baseDetails := SubscriptionDetails{
+var (
+	baseDetails = SubscriptionDetails{
 		MessageType: proto.MessageType_PREPARE,
 		View: &proto.View{
 			Height: 0,
 			Round:  0,
 		},
 		MinNumMessages: 1,
 	}
+)
+
+func TestEventManager_signalEvent(t *testing.T) {
+	t.Parallel()
+
+	var (
+		baseEventType = baseDetails.MessageType
+		baseEventView = &proto.View{
+			Height: baseDetails.View.Height,
+			Round:  baseDetails.View.Round,
+		}
+	)
+
+	// setupEventManagerAndSubscription creates new eventManager and a subscription
+	setupEventManagerAndSubscription := func(t *testing.T) (*eventManager, *Subscription) {
+		t.Helper()
+
+		em := newEventManager()
+		t.Cleanup(func() {
+			em.close()
+		})
+
+		subscription := em.subscribe(baseDetails)
+		t.Cleanup(func() {
+			em.cancelSubscription(subscription.ID)
+		})
+
+		return em, subscription
+	}
+
+	// emitEvent sends a event to eventManager and close doneCh after signalEvent completes
+	emitEvent := func(
+		t *testing.T,
+		em *eventManager,
+		eventType proto.MessageType,
+		eventView *proto.View,
+	) <-chan struct{} {
+		doneCh := make(chan struct{})
+
+		go func() {
+			t.Helper()
+			defer close(doneCh)
+
+			em.signalEvent(
+				eventType,
+				eventView,
+			)
+		}()
+
+		return doneCh
+	}
+
+	// testSubscriptionData checks the data sent to subscription
+	testSubscriptionData := func(
+		t *testing.T,
+		sub *Subscription,
+		expectedSignals []uint64,
+	) <-chan struct{} {
+		doneCh := make(chan struct{})
+
+		go func() {
+			t.Helper()
+			defer close(doneCh)
+
+			actualSignals := make([]uint64, 0)
+			for sig := range sub.SubCh {
+				actualSignals = append(actualSignals, sig)
+			}
+
+			assert.Equal(t, expectedSignals, actualSignals)
+		}()
+
+		return doneCh
+	}
+
+	// closeSubscription closes subscription manually
+	// because cancelSubscription might be unable
+	// due to mutex locking during tests
+	closeSubscription := func(
+		t *testing.T,
+		em *eventManager,
+		sub *Subscription,
+	) {
+		t.Helper()
+
+		close(em.subscriptions[sub.ID].doneCh)
+		delete(em.subscriptions, sub.ID)
+	}
+
+	t.Run("should exit before locking subscriptionsLock if numSubscriptions is zero", func(t *testing.T) {
+		t.Parallel()
+
+		em, sub := setupEventManagerAndSubscription(t)
+
+		// overwrite numSubscription
+		atomic.StoreInt64(&em.numSubscriptions, 0)
```

Comment:
Why do we need it? Could we do the same inside the guarded area(the lock one line below)?

Reply:
This is for safe check to make sure early return occurs if numSubscriptions is zero